testsuite/drivers/driver/python_script.py

```diff
@@ -60,3 +70,99 @@ def run(self):
         else:
             self.result.log += 'missing SUCCESS output line'
             raise TestAbortWithFailure('missing SUCCESS output line')
+
+
+    def save_working_dir(self):
+        # Save the working directory state for later restoration.
+
+        base = self.test_env['working_dir']
+        orig_name = ".orig"
+        orig = os.path.join(base, orig_name)
+
+        # Save the original files under ".orig" folder
+        os.mkdir(orig)
+        for f in os.listdir(base):
+            if f == orig_name:
+                continue
+            path = os.path.join(base, f)
+            if os.path.isfile(path):
+                shutil.copy(path, orig)
+            else:
+                shutil.copytree(path, os.path.join(orig, f))
+
+
+    def restore_working_dir(self):
+        # Restore the working directory to its initial state, by deleting
+        # everything and copying originals back from .orig dir
+
+        def make_writable(path):
+            # Make everything inside a directory writable recursively
+            for root, dirs, files in os.walk(path):
+                for d in dirs:
+                    os.chmod(os.path.join(root, d), 0o777)
+                for f in files:
+                    os.chmod(os.path.join(root, f), 0o666)
+
+        base = self.test_env['working_dir']
+        orig_name = ".orig"
+
+        # Delete anything not called ".orig"
+        for f in os.listdir(base):
+            if f != orig_name:
+                path = os.path.join(base, f)
+                if os.path.isfile(path):
+                    os.remove(path)
+                else:
+                    # Git marks some files read-only, so make them writable
+                    make_writable(path)
+                    shutil.rmtree(path)
+
+        # Restore the original files
+        orig = os.path.join(base, orig_name)
+        for f in os.listdir(orig):
+            path = os.path.join(orig, f)
+            if os.path.isfile(path):
+                shutil.copy(path, base)
+            else:
+                shutil.copytree(path, os.path.join(base, f))
+
+
+    def run(self):
+        # Run the test itself. Depending on the build mode, it may be run
+        # twice.
+        DEFAULT_MODE = "both"
+
+        pristine_env = self.prepare()
+
+        # Obtain the build mode for the test
+        mode = self.test_env.get('build_mode',
+                                 self.test_env.get('build-mode',
+                                                   DEFAULT_MODE))
+        # One of 'shared', 'sandboxed', or 'both'
+
+        # If mode is "both", track original files for later
+        if mode == "both":
+            self.save_working_dir()
```

Comment:
I feel like there must be a "better" way to this.

Maybe @pmderodat if you have a minute to have a look. We want to run all the tests in two different modes, is there a nice way to do this?

Reply:
While we wait for @pmderodat feedback (I also thought that this is hackish), I will merge provisionally, we can revert/update later if there is a better way.